### PR TITLE
Use Explore's Prometheus editor in dashboard panel edit

### PR DESCRIPTION
**Pull request by davkal**
- adds PromQueryField to query editor partial
- `query` is being pushed down for the initial value, and changes coming from the editor bubble up and override the panel target

### CODE REVIEW

File: packages/grafana-ui/src/types/plugin.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { ComponentClass } from 'react';
 import { PanelProps, PanelOptionsProps } from './panel';
 import { DataQueryOptions, DataQuery, DataQueryResponse, QueryHint, QueryFixAction } from './datasource';
+import { TimeRange } from './time';
+import { PanelModel } from 'app/features/dashboard/state';
```
Comment (torkelo) [inline packages/grafana-ui/src/types/plugin.ts]: You cannot import stuff from the app from grafana ui

File: packages/grafana-ui/src/types/plugin.ts [hunk 2]
```diff
@@ -49,7 +51,9 @@ export interface ExploreDataSourceApi<TQuery extends DataQuery = DataQuery> exte
 
 export interface QueryEditorProps<DSType extends DataSourceApi, TQuery extends DataQuery> {
   datasource: DSType;
+  panel?: PanelModel;
```
Comment (torkelo) [inline packages/grafana-ui/src/types/plugin.ts]: Don’t want plugins to have access to the panel model , is there a reason for this and why is it optional?

**Comment by torkelo** [comment]
Is this still WIP? what is the state? let me know if this can be tested and & merged. Or is it waiting for getting the latest query response added to the QueryEditorProps interface?

**Comment by davkal** [comment]
Basic functionality is done. Needs tests. Hope to finish off next week.

**Comment by davkal** [comment]
This is ready for an initial review.

Given that the editor component no longer has access to the panel controller, we're missing two things for the external link generation: 1) dynamic step interval from the panel, 2) template variables updates

The dynamic step interval is probably less important than the template variables. It's going to be very frustrating when someone exists to Prometheus and starts debugging, only to find out that they are operating on a different cluster/datacenter/instance. Any idea on how we can force a render of the link when the variables update? @torkelo

File: public/app/plugins/datasource/prometheus/components/PromQueryEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  onIntervalFactorChange = (option: SelectOptionItem) => {
+    this.query.intervalFactor = option.value;
+    this.setState({ intervalFactorOption: option }, this.onRunQuery);
+  };
+
+  onLegendChange = (e: React.SyntheticEvent<HTMLInputElement>) => {
+    const legendFormat = e.currentTarget.value;
+    this.query.legendFormat = legendFormat;
+    this.setState({ legendFormat });
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/components/PromQueryEditor.tsx]: should this no also trigger a run query so you see the effect?

File: public/app/plugins/datasource/prometheus/components/PromQueryEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+import { PrometheusDatasource } from '../datasource';
+import { PromQuery } from '../types';
+
+import PromQueryField from './PromQueryField';
+import PromLink from './PromLink';
+
+export type Props = QueryEditorProps<PrometheusDatasource, PromQuery>;
+
+const FORMAT_OPTIONS: SelectOptionItem[] = [
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/components/PromQueryEditor.tsx]: why caps?

**Comment by torkelo** [comment]
Did a quick poc for adding queryResponse & queryError to react query editor props: 
https://github.com/grafana/grafana/pull/16233

**Comment by davkal** [comment]
@torkelo any guidance on how panel editors should react to a user changing template variables? This is needed for the external link creation.

**Comment by torkelo** [comment]
We need pass variables in as props, but with the PR I linked to above you will also get a new query response as queries are executed in variable change, but to show variables in auto complete suggestions we need to pass variables in as props as well

**Comment by davkal** [comment]
@torkelo Great, I can trigger a render then. Will wait for #16233

**Comment by torkelo** [comment]
That PR (queryRespone & queryError) is now available in master, merged via this other PR from ryan https://github.com/grafana/grafana/pull/16306 

But added refId filtering to both queryError and queryResponse, so we need to add refId to errors and series responses in the Prometheus response handling.

**Comment by torkelo** [comment]
Now that we have queryResponse & queryError in the QueryEditorProps would be interesting as next steps to review how we can align & remove the Prometheus/loki specific from the Explore query interfaces:

```ts
export interface ExploreQueryFieldProps<DSType extends DataSourceApi, TQuery extends DataQuery> {
  datasource: DSType;
  query: TQuery;
  error?: string | JSX.Element;
  hint?: QueryHint;
  history: any[];
  onExecuteQuery?: () => void;
  onQueryChange?: (value: TQuery) => void;
  onExecuteHint?: (action: QueryFixAction) => void;
}
```

Would like to remove hint & onExecuteHint as they seem internal details that can be handled inside the QueryEditor. For example datasource.getQueryHints could be an internal Prometheus specific function   that the QueryEditor calls, does not need to be a public function on DataSourceApi.

**Comment by davkal** [comment]
LGTM. Thanks for taking this on @hugohaggmark 

IMO this can go in as is. Someone other than me needs to give a review (I opened the PR.)

The one piece of functionality we're missing over the existing editor is start/end awareness of metadata queries (when requesting labels, `start` and `end` from the visible range should be sent as well). We should add this in a follow-up PR.

**Review by hugohaggmark — APPROVED** [review]
LGTM!

**Pull request merged**